test/e2e/kcp_remediations.go

```diff
@@ -0,0 +1,697 @@
+	log.Logf("Sending bootstrap signal %s to Machine %s", input.Signal, input.Machine)
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapName,
+			Namespace: input.Namespace,
+		},
+	}
+	Expect(input.Client.Get(ctx, client.ObjectKeyFromObject(cm), cm)).To(Succeed(), "failed to get mhc-test config map")
+
+	cmWithSignal := cm.DeepCopy()
+	cmWithSignal.Data[configMapDataKey] = input.Signal
+	Expect(input.Client.Patch(ctx, cmWithSignal, client.MergeFrom(cm))).To(Succeed(), "failed to patch mhc-test config map")
+
+	log.Logf("Waiting for Machine %s to acknowledge signal %s has been received", input.Machine, input.Signal)
+	Eventually(func() string {
+		_ = input.Client.Get(ctx, client.ObjectKeyFromObject(cmWithSignal), cmWithSignal)
+		return cmWithSignal.Data[configMapDataKey]
+	}, "1m", "10s").Should(Equal(fmt.Sprintf("ack-%s", input.Signal)), "Failed to get ack signal from machine %s", input.Machine)
+
+	machine := &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      input.Machine,
+			Namespace: input.Namespace,
+		},
+	}
+	Expect(input.Client.Get(ctx, client.ObjectKeyFromObject(machine), machine)).To(Succeed())
+
+	// Resetting the signal in the config map
+	cmWithSignal.Data[configMapDataKey] = "hold"
+	Expect(input.Client.Patch(ctx, cmWithSignal, client.MergeFrom(cm))).To(Succeed(), "failed to patch mhc-test config map")
+}
+
+type waitForMachinesInput struct {
+	Lister                          framework.Lister
+	Namespace                       string
+	ClusterName                     string
+	ExpectedReplicas                int
+	ExpectedOldMachines             []string
+	ExpectedDeletedMachines         []string
+	WaitForMachinesIntervals        []interface{}
+	CheckMachineListStableIntervals []interface{}
+}
+
+// waitForMachines waits for machines to reach a well known state defined by number of replicas, a list of machines to exists,
+// a list of machines to not exist anymore. The func also check that the state is stable for some time before
+// returning the list of new machines.
+func waitForMachines(ctx context.Context, input waitForMachinesInput) (allMachineNames, newMachineNames []string) {
+	inClustersNamespaceListOption := client.InNamespace(input.Namespace)
+	matchClusterListOption := client.MatchingLabels{
+		clusterv1.ClusterNameLabel:         input.ClusterName,
+		clusterv1.MachineControlPlaneLabel: "",
+	}
+
+	expectedOldMachines := sets.Set[string]{}.Insert(input.ExpectedOldMachines...)
+	expectedDeletedMachines := sets.Set[string]{}.Insert(input.ExpectedDeletedMachines...)
+	allMachines := sets.Set[string]{}
+	newMachines := sets.Set[string]{}
+	machineList := &clusterv1.MachineList{}
+
+	// Waits for the desired set of machines to exist.
+	log.Logf("Waiting for %d machines, must have %s, must not have %s", input.ExpectedReplicas, expectedOldMachines.UnsortedList(), expectedDeletedMachines.UnsortedList())
+	Eventually(func() bool {
```

Comment:
If we did this instead we wouldn't swallow the individual errors:

```
Eventually(func(g Gomega) {
// Gets the list of machines
     err := input.Lister.List(ctx, machineList, inClustersNamespaceListOption, matchClusterListOption)
         g.Expect(err).NotTo(HaveOccurred())
...
...
...
}, input.WaitForMachinesIntervals...).Should(Succeed(), "Failed to get the expected list of machines: got %s (expected %d machines, must have %s, must not have %s)", allMachines.UnsortedList(), input.ExpectedReplicas, expectedOldMachines.UnsortedList(), expectedDeletedMachines.UnsortedList())
```

Basically if we pass in Gomega context to the eventually we can simply assert every condition that must be met in order to make forward progress, and for each violation we get a nice clean message with the short-circuit line of code, and for error types we get any strongly typed error debug data in the stdout/stderr message.

Reply:
Sounds good to me! I did something similar in a recent refatoring. Worked well